src/mbgl/style/expression/match.cpp

```diff
@@ -0,0 +1,240 @@
+    if (length < 5) {
+        ctx.error(
+            "Expected at least 4 arguments, but found only " + std::to_string(length - 1) + "."
+        );
+        return ParseResult();
+    }
+
+    // Expect odd-length array: ["match", input, 2 * (n pairs)..., otherwise]
+    if (length % 2 != 1) {
+        ctx.error("Expected an even number of arguments.");
+        return ParseResult();
+    }
+
+    optional<type::Type> inputType;
+    optional<type::Type> outputType;
+    if (ctx.expected && *ctx.expected != type::Value) {
+        outputType = ctx.expected;
+    }
+
+    std::vector<std::pair<std::vector<InputType>,
+                          std::unique_ptr<Expression>>> branches;
+
+    for (size_t i = 2; i + 1 < length; i += 2) {
+        const auto& label = arrayMember(value, i);
+
+        std::vector<InputType> labels;
+        // Match pair inputs are provided as either a literal value or a
+        // raw JSON array of string / number / boolean values.
+        if (isArray(label)) {
+            auto groupLength = arrayLength(label);
+            if (groupLength == 0) {
+                ctx.concat(i).error("Expected at least one branch label.");
+                return ParseResult();
+            }
+            
+            for (size_t j = 0; j < groupLength; j++) {
+                const optional<InputType> inputValue = parseInputValue(arrayMember(label, j), ctx.concat(i, inputType), inputType);
+                if (!inputValue) {
+                    return ParseResult();
+                }
+                labels.push_back(*inputValue);
+            }
+        } else {
+            const optional<InputType> inputValue = parseInputValue(label, ctx.concat(i, inputType), inputType);
+            if (!inputValue) {
+                return ParseResult();
+            }
+            labels.push_back(*inputValue);
+        }
+        
+        ParseResult output = ctx.concat(i + 1, outputType).parse(arrayMember(value, i + 1));
+        if (!output) {
+            return ParseResult();
+        }
+        
+        if (!outputType) {
+            outputType = (*output)->getType();
+        }
+        
+        branches.push_back(std::make_pair(std::move(labels), std::move(*output)));
+    }
+
+    auto input = ctx.concat(1, inputType).parse(arrayMember(value, 1));
+    if (!input) {
+        return ParseResult();
+    }
+
+    auto otherwise = ctx.concat(length - 1, outputType).parse(arrayMember(value, length - 1));
+    if (!otherwise) {
+        return ParseResult();
+    }
+
+    assert(inputType && outputType);
+
+    return inputType->match(
+        [&](const type::NumberType&) {
+            return create<int64_t>(*outputType, std::move(*input), std::move(branches), std::move(*otherwise), ctx);
+        },
+        [&](const type::StringType&) {
+            return create<std::string>(*outputType, std::move(*input), std::move(branches), std::move(*otherwise), ctx);
+        },
+        [&](const auto&) {
+            assert(false);
```

Comment:
Why can this never be reached?

Reply:
inputType is set by parseInputValue(), which only accepts string and (integer) numeric values.  I'll add a comment.